OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/ifds/VariableTypeProblem.scala

```diff
@@ -0,0 +1,357 @@
+import org.opalj.collection.immutable.EmptyIntTrieSet
+import org.opalj.ifds.AbstractIFDSFact
+
+import org.opalj.tac.fpcf.analyses.ifds.{JavaIFDSProblem => NewJavaIFDSProblem}
+import org.opalj.tac._
+import org.opalj.value.ValueInformation
+import scala.annotation.tailrec
+
+import org.opalj.ifds.AbstractIFDSNullFact
+import org.opalj.ifds.Dependees.Getter
+
+import org.opalj.fpcf.FinalEP
+import org.opalj.br.analyses.DeclaredMethodsKey
+import org.opalj.br.fpcf.PropertyStoreKey
+import org.opalj.tac.fpcf.properties.cg.Callers
+
+trait VTAFact extends AbstractIFDSFact
+case object VTANullFact extends VTAFact with AbstractIFDSNullFact
+
+/**
+ * A possible run time type of a variable.
+ *
+ * @param definedBy The variable's definition site.
+ * @param t The variable's type.
+ * @param upperBound True, if the variable's type could also be every subtype of `t`.
+ */
+case class VariableType(definedBy: Int, t: ReferenceType, upperBound: Boolean) extends VTAFact {
+
+    /**
+     * If this VariableType is an upper bound, it subsumes every subtype.
+     */
+    override def subsumes(other: AbstractIFDSFact, project: SomeProject): Boolean = {
+        if (upperBound) other match {
+            case VariableType(definedByOther, tOther, _) if definedBy == definedByOther && t.isObjectType && tOther.isObjectType =>
+                project.classHierarchy.isSubtypeOf(tOther.asObjectType, t.asObjectType)
+            case _ => false
+        }
+        else false
+    }
+}
+
+/**
+ * A possible run time type of the receiver of a call.
+ *
+ * @param line The line of the call.
+ * @param t The callee's type.
+ * @param upperBound True, if the callee's type could also be every subtype of `t`.
+ */
+case class CalleeType(line: Int, t: ReferenceType, upperBound: Boolean) extends VTAFact {
+
+    /**
+     * If this CalleeType is an upper bound, it subsumes every subtype.
+     */
+    override def subsumes(other: AbstractIFDSFact, project: SomeProject): Boolean = {
+        if (upperBound) other match {
+            case CalleeType(lineOther, tOther, _) if line == lineOther && t.isObjectType && tOther.isObjectType =>
+                tOther.asObjectType.isSubtypeOf(t.asObjectType)(project.classHierarchy)
+            case _ => false
+        }
+        else false
+    }
+}
+
+class VariableTypeProblem(project: SomeProject, override val subsumeFacts: Boolean = false) extends JavaIFDSProblem[VTAFact](project) {
+    val propertyStore = project.get(PropertyStoreKey)
+    val declaredMethods = project.get(DeclaredMethodsKey)
+
+    override def nullFact: VTAFact = VTANullFact
+
+    /**
+     * The analysis starts with all public methods in java.lang or org.opalj.
+     */
+    override def entryPoints: Seq[(Method, VTAFact)] = {
+        project.allProjectClassFiles
+            .filter(classInsideAnalysisContext)
+            .flatMap(classFile => classFile.methods)
+            .filter(isEntryPoint)
+            .flatMap(entryPointsForMethod)
```

Comment:
While probably not critical, this looks a like efficiency could be improved here